### PR TITLE
DAOS-15076 engine: Fixes and Performance improvements 

### PR DESCRIPTION
**Pull request by wangshilong**
1. Set @sra_flags after sched_req_attr_init()
2. Moved req_need_reject() to an earlier position.
3. Reduced size of req_comm_in by 32 bytes.
5. Resolved interoperability issues between new clients and old servers.
6. Implementing a walkaround to avoid enqueuing metadata RPCS can
improve stat and removal operations by 2-3%.

Required-githooks: true

### CODE REVIEW

File: src/include/daos/rpc.h [hunk 1]
```diff
@@ -126,6 +126,8 @@ struct daos_req_comm_in {
 	uint32_t	req_in_gid;
 	/** Request project ID, reserved for NRS */
 	uint32_t	req_in_projid;
+	/* padding reserved */
+	uint32_t	req_in_padding0;
```
[the remark this answers is not in the document]
Comment (wangshilong) [reply src/include/daos/rpc.h:130]: Reduce structure size is still helpful, i will reduce this struct size.

File: src/engine/sched.c [hunk 2]
```diff
@@ -1305,7 +1308,24 @@ should_enqueue_req(struct dss_xstream *dx, struct sched_req_attr *attr)
...
+	 * have indicated a 2%~3% drop in stat and removal operations when
+	 * this is done. It may be worthwhile to reassess this decision in
+	 * the future, especially if Quality of Service(QoS) requirements
+	 * are introduced. (See DAOS-15076)
+	 */
+	ret = ABT_pool_get_size(abt_pool, &cnt);
+	if (ret == ABT_SUCCESS && cnt < MAX_KICKED_REQ_CNT)
+		return false;
+
+	return true;
```
[the remark this answers is not in the document]
Comment (wangshilong) [reply src/engine/sched.c:1328]: OK.

File: src/object/srv_mod.c [hunk 3]
```diff
@@ -210,13 +203,15 @@ obj_get_req_attr(crt_rpc_t *rpc, struct sched_req_attr *attr)
 		sched_req_attr_init(attr, obj_rpc_is_update(rpc) ?
 				    SCHED_REQ_UPDATE : SCHED_REQ_FETCH,
 				    &orw->orw_pool_uuid);
+		attr->sra_flags |= SCHED_REQ_FL_FORCE_ENQUEUE;
```
[the remark this answers is not in the document]
Comment (wangshilong) [reply src/object/srv_mod.c:206]: I thought that we still need enqueue these metadata RPCs if system is overloaded, As discussed i am ok to walkaround this issue to leave it as SCHED_REQ_ANONYM.

File: src/object/srv_mod.c [hunk 4]
```diff
@@ -279,41 +279,41 @@ obj_get_req_attr(crt_rpc_t *rpc, struct sched_req_attr *attr)
...
-		sched_req_attr_init(attr, SCHED_REQ_FETCH, &oki->oki_pool_uuid);
+		sched_req_attr_init(attr, SCHED_REQ_ANONYM, &oki->oki_pool_uuid);
 		break;
 	}
 	case DAOS_OBJ_RPC_EC_AGGREGATE: {
 		struct obj_ec_agg_in *ea = crt_req_get(rpc);
 
 		attr->sra_enqueue_id = ea->ea_comm_in.req_in_enqueue_id;
-		sched_req_attr_init(attr, SCHED_REQ_MIGRATE, &ea->ea_pool_uuid);
+		sched_req_attr_init(attr, SCHED_REQ_UPDATE, &ea->ea_pool_uuid);
```
Comment (NiuYawei) [inline src/object/srv_mod.c:289]: ANONYM?

File: src/object/srv_mod.c [hunk 5]
```diff
@@ -219,6 +212,13 @@ obj_get_req_attr(crt_rpc_t *rpc, struct sched_req_attr *attr)
 		sched_req_attr_init(attr, SCHED_REQ_MIGRATE, &omi->om_pool_uuid);
 		break;
 	}
+	/*
+	 * To enhance system performance, metadata RPCs are currently not
```
Comment (NiuYawei) [inline src/object/srv_mod.c:216]: 'metadata RPCs' looks confusing to me.

**Review by NiuYawei — DISMISSED** [review]

File: src/object/srv_mod.c (hunk 5 above)
Comment (wangshilong) [reply src/object/srv_mod.c]: Indeed.

File: src/object/srv_mod.c (hunk 4 above)
Comment (wangshilong) [reply src/object/srv_mod.c]: fixed.

**Review by NiuYawei — APPROVED** [review]

**Review by Nasf-Fan — APPROVED** [review]

File: src/include/daos/rpc.h [hunk 6]
```diff
@@ -120,20 +120,10 @@ enum daos_rpc_type {
 };
 
 struct daos_req_comm_in {
-	/** Request user ID, reserved for NRS */
-	uint32_t	req_in_uid;
-	/** Request group ID, reserved for NRS */
-	uint32_t	req_in_gid;
```
Comment (liuxuezhao) [inline src/include/daos/rpc.h:126]: why remove these fields? plan to use the req_in_paddings in future?

File: src/object/obj_rpc.c [hunk 7]
```diff
@@ -1072,12 +1063,6 @@ crt_proc_struct_daos_req_comm_in(crt_proc_t proc, crt_proc_op_t proc_op,
 		if (rc)
 			return rc;
 	}
```
Comment (liuxuezhao) [inline src/object/obj_rpc.c:1065]: rc = crt_proc_memcpy(proc, proc_op, drci, sizeof(*drci));
this one looks enough, need not pack it one by one.

**Review by liuxuezhao — APPROVED** [review]

File: src/include/daos/rpc.h (hunk 6 above)
Comment (wangshilong) [reply src/include/daos/rpc.h:126]: Yes.

**Pull request merged**